### ISSUE TITLE
Fine tuning the network for new data.

### ISSUE THREAD

**Issue by pushkarjajoria** [schufo/lyrics-aligner#2]
Hey @schufo ,

I am a research assistant at the University of Galway, Ireland. I am working on lyrics alignment for Italian Arias. As expected the model is not great at generalizing to the new style of music (operas) and the alignment is a bit off. My plan is to fine-tune the network with existing weights on the new dataset. 
While going through the code, I didn't find a train.py (or a training loop). Just so you know, I cannot train the whole network(i.e lyrics aligner + voice separator) because I do not have the training data for separated voices. It'd be great if you could point me in the right direction regarding this so I can write the training loop or use an existing training loop with new data.

-
Pushkar

**Comment by schufo** [comment]
Hi Pushkar,

The training script for the model can be found in this repo: https://github.com/schufo/plla-tisvs

You could ignore the part of the model which predicts the solo singing voice (only use the model up to the attention mechanism) and fine tune it with a loss function on the attention matrix. I think a loss function for this purpose was proposed in this paper:

Nishikimi, R., Nakamura, E., Fukayama, S., Goto, M., & Yoshii, K. (2019, May). Automatic singing transcription based on encoder-decoder recurrent neural networks with a weakly-supervised attention mechanism. In ICASSP 2019-2019 IEEE International Conference on Acoustics, Speech and Signal Processing (ICASSP) (pp. 161-165)

I would also like to mention that my former colleague Charles worked on lyrics alignment for opera. Maybe you find something useful in his work: https://scholar.google.com/citations?user=kPKpcyAAAAAJ&hl=de&oi=ao

I hope this helps :)

**Comment by pushkarjajoria** [comment]
Hey @schufo ,

Thanks a lot for answering. It make sense to me. I can just comment out the part of the model that is doing voice separation. For instance, talking about the `InformedOpenUnmix3` model, if I remember correctly, `alphas` contain aligned timestamps for phonemes. Anything beyond alphas would be ignored/commented.

Regarding the Loss function: In your code, you are using the L1Loss whereas the loss function in the paper that you have mentioned is different. Did you also use the loss function that is mentioned in the paper or is it something nice to do but you trained the model without it? 

I have another heuristic in mind and it'd be really helpful if you can share your insights about that. 
The usecase for which I am training the model only requires that I provide correct word-level timestamps (i.e. the start phoneme and the word end symbol). Creating a dataset of phoneme-level timestamps is tedious work and we have decided to work at word level instead. My plan is to only compute the loss for the word start and end timestamps. I'll just ignore the intermediary phonemes(the intermediary timestamps for training data are computed by placing them uniformly between the start and end).  Any thoughts about this?

Thanks again for all the help and Thanks for sharing the papers from Charles! They do seem relevant.

**Comment by schufo** [comment]
Regarding the loss function: I trained the model with the objective to predict the singing voice (voice separation) and the alignment was learned through the attention weights as a side effect. Since you don't have data to train the model for separation, you have to use a loss function on the alignment quality directly. That's why I mentioned the paper.

I am not sure if training the model with time stamps at word level will work. Maybe in this case you need to look for another model. 

You can look here for some related work:
https://scholar.google.com/citations?hl=de&user=rw6W7WYAAAAJ

https://scholar.google.com/citations?user=NFi7pkcAAAAJ&hl=de&oi=sra

McAuliffe, M., Socolof, M., Mihuc, S., Wagner, M., & Sonderegger, M. (2017, August). Montreal forced aligner: Trainable text-speech alignment using kaldi. In Interspeech (Vol. 2017, pp. 498-502).

**Comment by pushkarjajoria** [comment]
Got it. Thanks for the information.
